### PR TITLE
Aztec - Fix problems with media thumbnails that were loaded at different sizes

### LINKED ISSUE #6945: Aztec Media Placeholders - Make sure thumbnails are resized accordingly to the editor viewport size
We should make sure that Media added to Aztec are properly resized to fit the editor viewport size. 

Also, we should make sure the same size is used during the uploading process, and failures. See #6942
device-2017-12-05-154554

At the moment picture have different sizes when uploaded from the device, or opened a pre-existing post.

Note 1: When the device is rotated we should resize them as well.
Note 2: Make sure to not upsample pictures.

### PR DESCRIPTION
**Pull request by daniloercoli**
Fixes #6945 (and partially #6942 that requires an Aztec patch for full resolution), by making sure pictures added to the editor are properly resized by the MediaLoader(s) to the correct width requested by the editor. The max width for pictures is set by us at run-time by using the value returned in `ImageUtils.getMaximumThumbnailWidthForEditor(getActivity())`.

One of the problem fixed was different thumb sizes for the same picture in the editor. The culprit  is the way the preview thumb is created by the functions that load it from the disk, or from the network. Volley code does only take in consideration the width of the picture, instead our ImageUtils does take in consideration both height and width of the picture. So the different sizes of the preview thumb for portrait pictures if loaded from the network.

Also fixed the problem with placeholders that could appear too small in some cases. 
_Note that for a full resolution of this, error placeholder that appear when a network error occurs,  we need a patch in Aztec._

Test 1:
- Add pictures to the editor. 
- Wait until both of those are uploaded
- Exit the editor
- Re-open the editor and pictures should have the same size

Test 2:
- Add pictures to the editor from local device. 
- Add picture to the editor from Media Library
- All the pictures should have the same size (* make sure to use pictures bigger than the max size, since pictures are not upscaled. If you use small pictures than their real size is used).

cc  @mzorz

### CODE REVIEW

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java [hunk 1]
```diff
@@ -181,15 +185,21 @@ public void onCreate(Bundle savedInstanceState) {
...
-            ((EditorFragmentActivity) getActivity()).initializeEditorFragment();
-        }
-
         title = (EditTextWithKeyBackListener) view.findViewById(R.id.title);
         content = (AztecText) view.findViewById(R.id.aztec);
         source = (SourceViewEditText) view.findViewById(R.id.source);
 
+        // Set the default value for max and min picture sizes.
+        maxImageWidthForVisualEditor = ImageUtils.getMaximumThumbnailWidthForEditor(getActivity());
+        minImageWidthForVisualEditor = DisplayUtils.dpToPx(getActivity(), MIN_BITMAP_DIMENSION_DP);
```
Comment (mzorz) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: This is a nitpick and not direct part of the changes here, but looks relevant: what do you think about modifying `ImageUtils.getMaximumThumbnailWidthForEditor() ` to receive `MIN_BITMAP_DIMENSION_DP` as a parameter along with the `Context`, and use it instead the `48` hardcoded value in there?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 2]
```diff
@@ -1041,17 +1041,23 @@ public void initializeEditorFragment() {
             aztecEditorFragment.setEditorBetaClickListener(EditPostActivity.this);
             aztecEditorFragment.setEditorImageSettingsListener(EditPostActivity.this);
 
-            Drawable loadingImagePlaceholder = getResources().getDrawable(org.wordpress.android.editor.R.drawable.ic_gridicons_image);
-            loadingImagePlaceholder.setBounds(0, 0,
-                    AztecEditorFragment.DEFAULT_MEDIA_PLACEHOLDER_DIMENSION_DP,
-                    AztecEditorFragment.DEFAULT_MEDIA_PLACEHOLDER_DIMENSION_DP);
+            // Here we should set the max width for picture, but the default size is already OK. No need
+            // to customize it further
+            // int maxImageWidthForVisualEditor = ImageUtils.getMaximumThumbnailWidthForEditor(getActivity());
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: we can remove this commented code line

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/MediaUtils.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            Canvas canvas = new Canvas(bitmap);
+            drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
+            drawable.draw(canvas);
+        } else {
+            throw new IllegalArgumentException("Unsupported drawable type");
+        }
+        bitmap.setDensity(DisplayMetrics.DENSITY_DEFAULT);
+        return new BitmapDrawable(context.getResources(), bitmap);
+    }
+}
```
Comment (mzorz) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/MediaUtils.java]: What do you think about moving `getMaximumThumbnailWidthForEditor()` from the `ImageUtils` class to this new editor-focused class in  `/android/editor/MediaUtils.java`? Seems like `getMaximumThumbnailWidthForEditor` is very specific to the Editor, so it looks like that method now belongs  more to this file here than the pre-existing`ImageUtils` class

**Comment by daniloercoli** [comment]
Hey @mzorz , I was reading your comment in my other Aztec PR, and wondering if we should take in consideration both the width and the height of the media when creating the thumbnail. 
Otherwise in landscape portrait media could appear with a giant height. (Basically don't do this https://github.com/wordpress-mobile/WordPress-Android/pull/7057/files#diff-bc0574560dbbe433195cd3374ec2732fR34), and make the video/image Volley loaders return a media with both height and width that are less than our `editor maximum size`.  All of our `ImageUtils` methods already do this, so it should even easier for us to respect that policy.

Let me know what do you think, so I will patch this PR soon, and avoid to add other methods/complexity to image on Aztec. Also it will be in par with the current Visual Editor.

**Comment by mzorz** [comment]
> wondering if we should take in consideration both the width and the height of the media when creating the thumbnail.
> and make the video/image Volley loaders return a media with both height and width that are less than our editor maximum size

I think this is the right way as well! Let's go ahead with your proposal 👍

**Comment by daniloercoli** [comment]
Mario, this is now ready for another pass.

I've slightly changed the code so it does respect `maxSize` parameter for both  `w` and `h` when resize a picture. That's is basically what our ImageUtils methods already do internally. With this change our editor code is much cleaner than before.
_Note that now, when in landscape, portrait pictures don't expand to full width as expected._

**Comment by mzorz** [comment]
Re: 
> I was reading your comment in my other Aztec PR, and wondering if we should take in consideration both the width and the height of the media when creating the thumbnail.

Just leaving a link here for reference https://github.com/wordpress-mobile/AztecEditor-Android/pull/593#issuecomment-355757696

> Mario, this is now ready for another pass.

Looking at it now 👍

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 4]
```diff
@@ -1482,7 +1481,7 @@ public int getCount() {
 
     private int getMaximumThumbnailWidthForEditor() {
         if (mMaxThumbWidth == 0) {
-            mMaxThumbWidth = ImageUtils.getMaximumThumbnailWidthForEditor(this);
+            mMaxThumbWidth = org.wordpress.android.editor.MediaUtils.getMaximumThumbnailSizeForEditor(this);
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: I was wondering, maybe we can rename this class from `MediaUtils` to a prefixed name like`EditorMediaUtils`, even when that would be redundant with the `org.wordpress.android.editor` path, but just to be able to make a clearer distinction between all the `MediaUtils` we are using in this file. We have:
- `import org.wordpress.android.util.MediaUtils;`
- `import org.wordpress.android.util.WPMediaUtils;`
- and `org.wordpress.android.editor.MediaUtils` used in this line here

**Comment by mzorz** [comment]
Tested, works 👌 
Just one [minor change](https://github.com/wordpress-mobile/WordPress-Android/pull/7057/files#r160124415) before merging and we're ready to go

**Comment by mzorz** [comment]
While testing this a bit more, and following the testing steps described in the issue #6492, I found this behavior:

https://cloudup.com/cBVTkqLDR70

The logs show this:
```
01-08 09:20:29.391 27168-27285/org.wordpress.android E/Volley: [1202] BasicNetwork.performRequest: Unexpected response code 404 for https://testmzorz2.files.wordpress.com/2018/01/vid-20180107-wa0010_std.original.jpg
```
(basically, the loading placeholder is shown, and the network request throws 404 so the error icon is shown). Whatever the reasons for the thumbnail URL to be throwing 404 are, the change in icon size seems to be due to the error icon to be 48x48, while the loading icon size is calculated from Aztec's min size as per this PR. Should we keep the error icon the same size as the placeholder?

**Comment by daniloercoli** [comment]
Renamed `MediaUtils`.

>  Whatever the reasons for the thumbnail URL to be throwing 404 are, the change in icon size seems to be due to the error icon to be 48x48, while the loading icon size is calculated from Aztec's min size as per this PR. Should we keep the error icon the same size as the placeholder?

It should be of the same size. As long as Aztec reference is update to the last hash.

**Comment by mzorz** [comment]
LGTM :shipit:

**Comment by mzorz** [comment]
> the change in icon size seems to be due to the error icon to be 48x48, while the loading icon size is calculated from Aztec's min size as per this PR. Should we keep the error icon the same size as the placeholder?

This is addressed by PR https://github.com/wordpress-mobile/AztecEditor-Android/pull/597, merging and closing this one as it's ready now 👍 

LGTM :shipit:

**Pull request merged**